File: javascript/stringtools/machine.js
// emitted from framec_v0.10.0

const FrameEvent = require("./framelang/framelang");

// get include files at https://github.com/frame-lang/frame-ancillary-files

class StringTools {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sRouter_;
        this.#compartment = new StringToolsCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    reverse(str) {
        let e = FrameEvent("reverse",{"str":str});
        this.#mux(e);
        return e._return;
    }
    
    makePalindrome(str) {
        let e = FrameEvent("makePalindrome",{"str":str});
        this.#mux(e);
        return e._return;
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sRouter_:
                this.#sRouter_(e);
                break;
            case this.#sReverse_:
                this.#sReverse_(e);
                break;
            case this.#sMakePalindrome_:
                this.#sMakePalindrome_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sRouter_(e) {
        switch (e._message) {
            case "makePalindrome":
                {
                // make\npalindrome
                let compartment =  new StringToolsCompartment(this.#sMakePalindrome_);
                
                compartment._forwardEvent = e;
                
                this.#transition(compartment);
                return;
                }
                
            case "reverse":
                {
                // reverse
                let compartment =  new StringToolsCompartment(this.#sReverse_);
                
                compartment._forwardEvent = e;
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    #sReverse_(e) {
        switch (e._message) {
            case "reverse":
                {
                e._return = this.reverse_str_do((e._parameters["str"]));
                // ready
                let compartment =  new StringToolsCompartment(this.#sRouter_);
                
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    #sMakePalindrome_(e) {
        switch (e._message) {
            case "makePalindrome":
                {
                e._return = (e._parameters["str"]) + this.reverse_str_do((e._parameters["str"]));
                // ready
                let compartment =  new StringToolsCompartment(this.#sRouter_);
                
                
                this.#transition(compartment);
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    reverse_str_do(str) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    
    
};


//=============== Compartment ==============//

class StringToolsCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


// /********************

class StringToolsController extends StringTools {

	constructor() {
	  super()
	}
	reverse_str_do(str) {
		let result = ""
		for (let char of str) {
			result = char + result
		}
		return result
	}
};

module.exports = StringToolsController
// ********************/

File: python/stringtools/machine.py
# emitted from framec_v0.10.0
# get include files at https://github.com/frame-lang/frame-ancillary-files
from framelang.framelang import FrameEvent

class StringTools:
    
    def __init__(self):
        
        # Create and intialize start state compartment.
        self.__state = self.__stringtools_state_Router
        self.__compartment: 'StringToolsCompartment' = StringToolsCompartment(self.__state)
        self.__next_compartment: 'StringToolsCompartment' = None
        
        # Initialize domain
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__mux(frame_event)
    
    # ===================== Interface Block =================== #
    
    def reverse(self,str: str):
        parameters = {}
        parameters["str"] = str

        e = FrameEvent("reverse",parameters)
        self.__mux(e)
        return e._return
    
    def makePalindrome(self,str: str):
        parameters = {}
        parameters["str"] = str

        e = FrameEvent("makePalindrome",parameters)
        self.__mux(e)
        return e._return
    
    # ====================== Multiplexer ==================== #
    
    def __mux(self, e):
        if self.__compartment.state == self.__stringtools_state_Router:
            self.__stringtools_state_Router(e)
        elif self.__compartment.state == self.__stringtools_state_Reverse:
            self.__stringtools_state_Reverse(e)
        elif self.__compartment.state == self.__stringtools_state_MakePalindrome:
            self.__stringtools_state_MakePalindrome(e)
        
        if self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            if(next_compartment.forward_event is not None and 
               next_compartment.forward_event._message == ">"):
                self.__mux(FrameEvent( "<", self.__compartment.exit_args))
                self.__compartment = next_compartment
                self.__mux(next_compartment.forward_event)
            else:
                self.__do_transition(next_compartment)
                if next_compartment.forward_event is not None:
                    self.__mux(next_compartment.forward_event)
            next_compartment.forward_event = None
    
    # ===================== Machine Block =================== #
    
    def __stringtools_state_Router(self, e):
        if e._message == "makePalindrome":
            # make\npalindrome
            compartment = StringToolsCompartment(self.__stringtools_state_MakePalindrome)
            compartment.forward_event = e
            self.__transition(compartment)
            return
        
        elif e._message == "reverse":
            # reverse
            compartment = StringToolsCompartment(self.__stringtools_state_Reverse)
            compartment.forward_event = e
            self.__transition(compartment)
            return
        
    def __stringtools_state_Reverse(self, e):
        if e._message == "reverse":
            e._return = self.reverse_str_do(e._parameters["str"])
            # ready
            compartment = StringToolsCompartment(self.__stringtools_state_Router)
            self.__transition(compartment)
            return
        
    def __stringtools_state_MakePalindrome(self, e):
        if e._message == "makePalindrome":
            e._return = e._parameters["str"] + self.reverse_str_do(e._parameters["str"])
            # ready
            compartment = StringToolsCompartment(self.__stringtools_state_Router)
            self.__transition(compartment)
            return
        
    
    # ===================== Actions Block =================== #
    
    
    # Unimplemented Actions
    
    def reverse_str_do(self,str: str):
        raise NotImplementedError
    
    
    # =============== Machinery and Mechanisms ============== #
    
    def __transition(self, compartment: 'StringToolsCompartment'):
        self.__next_compartment = compartment
    
    def  __do_transition(self, next_compartment: 'StringToolsCompartment'):
        self.__mux(FrameEvent("<", self.__compartment.exit_args))
        self.__compartment = next_compartment
        self.__mux(FrameEvent(">", self.__compartment.enter_args))
    

# ===================== Compartment =================== #

class StringToolsCompartment:

    def __init__(self, state):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = FrameEvent(None, None)
    


# ********************

#class StringToolsController(StringTools):
	#def __init__(self,):
	    #super().__init__()

    #def reverse_str_do(self,str: str):
        #pass

# ********************


